### DISCUSSION TITLE
[FR] Copy toolbars/items and Import/Export toolbars

### DISCUSSION THREAD

**Discussion by Moyf** [chrisgurney/obsidian-note-toolbar#113] [Ideas]
Sometimes I have several similar items, with only minor changes between them. 
It would be beneficial to have some utility buttons to streamline this process! 
![image](https://github.com/user-attachments/assets/9b3f1755-6920-4cac-9581-bc7cbd5b0dfe)

(Although, for now, it doesn't take too much time to manage. :D)

---

BTW, do you think it would be necessary to add an **import/export** feature for the toolbar? 
![image](https://github.com/user-attachments/assets/698c2076-e3da-46f3-98c1-365987933441)

For instance, after recommending this plugin to my friends, I could share some commonly useful toolbar presets with them to help them get started quickly!

However, I understand that the file paths and commands may vary significantly, which could make sharing less meaningful. 

——Unless the commands are all native to OB, or if presets (icons, text) are provided, allowing users to only modify the final data? 🤔

**Comment by chrisgurney** [comment]
Some of this stuff is actually next on my list! More to come. 😄

**Reply by MrComputron** [reply]
Fantastic news.

**Comment by chrisgurney** [comment]
Duplicate (partially) of #48

**Comment by chrisgurney** [comment]
@Moyf 

> However, I understand that the file paths and commands may vary significantly, which could make sharing less meaningful.
> ——Unless the commands are all native to OB, or if presets (icons, text) are provided, allowing users to only modify the final data?

Regarding commands, what I'm thinking is that I can parse out the toolbar's ID from the command, and provide a link to Community Plugins to (optionally) install it. I'll attempt to update the Settings UI to highlight missing plugins in this way.

Your comment reminds me that I'll need to elevate the reporting of errors so they're easier to spot, and include "missing" files/folders as well.

I've been looking at how iOS/Mac Shortcuts works when sharing shortcuts with other users, which:
- highlights steps in the process when apps the shortcut depends on is missing;
- allows you do to define [import questions](https://support.apple.com/en-ca/guide/shortcuts/apdf330fd3a0/7.0/ios/17.0) when sharing a shortcut, to fill in variables.

**Reply by Moyf** [reply]
> @Moyf
> 
> > However, I understand that the file paths and commands may vary significantly, which could make sharing less meaningful.
> > ——Unless the commands are all native to OB, or if presets (icons, text) are provided, allowing users to only modify the final data?
> 
> Regarding commands, what I'm thinking is that I can parse out the toolbar's ID from the command, and provide a link to Community Plugins to (optionally) install it. I'll attempt to update the Settings UI to highlight missing plugins in this way.
> 
> Your comment reminds me that I'll need to elevate the reporting of errors so they're easier to spot, and include "missing" files/folders as well.
> 
> I've been looking at how iOS/Mac Shortcuts works when sharing shortcuts with other users, which:
> 
> * highlights steps in the process when apps the shortcut depends on is missing;
> * allows you do to define [import questions](https://support.apple.com/en-ca/guide/shortcuts/apdf330fd3a0/7.0/ios/17.0) when sharing a shortcut, to fill in variables.

Oh yes, considering these aspects when sharing is definitely necessary!

Compared to installation, "opening the corresponding plugin page" might be less "intrusive," which is more user-friendly.

> ![image](https://github.com/user-attachments/assets/601308df-176b-4771-b6ae-3a46f5145ef3)
> (Just like the Install button on [this website](https://obsidian.md/plugins?search=note%20toolbar#) does!)

---

Additionally, highlighting "missing content" (while retaining the possibility of reuse) would provide a great user experience. 
—Especially compared to "popping up an error warning for missing xxx" :P

Import issues are also a great idea! 
Besides that, there could be simpler "import instructions," just to inform users about "what this toolbar is for... it may depend on the xx plugin..." and other basic information.

**Reply by chrisgurney** [reply]
> Compared to installation, "opening the corresponding plugin page" might be less "intrusive," which is more user-friendly.

Yup, that's exactly was I was thinking -- i.e., just link to the URL on obsidian.md and let the user read about the plugin and decide what to do next.

> Besides that, there could be simpler "import instructions," just to inform users about "what this toolbar is for... it may depend on the xx plugin..." and other basic information.

I thought about that. But I was thinking about how Dataview and Templater have a huge fan base, and the majority of sharing of ideas and scripts are just copy/pasted into forum and chat messages.

As a result, I was thinking that import/export would really just have to produce and consume callout markdown. Then people can talk about what each toolbar is for in a very flexible (and familiar) way (as well as have a working toolbar in one less step, if they choose to have it in their note).

While callouts now support menus, commands, and folders, there's _some_ elements that aren't supported, such as visibility and tooltips... but I don't believe that's a deal breaker here. (Perhaps an option later could be to export as HTML, but that would complicate an understanding of what the toolbar does.)

As for icons, those could be supported with the [Iconize](https://github.com/FlorianWoelki/obsidian-iconize) plugin (and presumably others), which I'm thinking I'll have indirect support for via a user-defined templating mechanism, e.g., `:Lu{{ICON_NAME}}:`. Here's how Iconize works with callouts today:

![CleanShot 2024-07-19 at 08 34 25](https://github.com/user-attachments/assets/6d886188-ed6d-42cb-870c-3a85f126a25a)

**Comment by chrisgurney** [comment]
@Moyf @MrComputron FYI the ability to duplicate toolbars and toolbar items is now available in the [1.11.0 release](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.11.0).

**Reply by Moyf** [reply]
> @Moyf @MrComputron FYI the ability to duplicate toolbars and toolbar items is now available in the [1.11.0 release](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.11.0).

Great!

**Comment by chrisgurney** [comment]
@Moyf FYI, exporting toolbars to markdown is now available [in this beta](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.14-beta-2). Check out [the documentation](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Importing-and-exporting). Feedback welcome!

**Comment by chrisgurney** [comment]
@Moyf FYI, you can now share toolbars  [in the beta](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/153) by right-clicking on a toolbar and selecting:

- _Share toolbar_, which gives you a URI you can send to another user (once they have the latest release installed).
- _Copy as markdown_, gives you Note Toolbar Callout markdown. Icons in callouts require the Iconize plugin to be installed + enabled.

I'd love feedback on this! I've still got some UX polish to do, but it works at the moment!
